File: 백준/Gold/21772. 가희의 고구마 먹방/가희의 고구마 먹방.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		T = Integer.parseInt(st.nextToken());
		
		visited = new boolean[R+1][C+1];
		board = new char[R+1][C+1];
		int sy = 0;
		int sx = 0;
		for(int y=1; y<=R; y++) {
			String line = br.readLine();
			for(int x=1; x<=C;x++) {
				board[y][x] = line.charAt(x-1);
				if(board[y][x] == 'G') {
					sy = y;
					sx = x;
				}
					
			}
		}
		visited[sy][sx] = true;
		System.out.println(solve(sy,sx,0,0));
		
	}
	
	static int solve(int y,int x,int cnt ,int t) {
		if(t == T)
			return cnt;
		
		int ret = solve(y,x,cnt , t+1 );
		for(int dir = 0; dir< 4; dir++) {
			int ny = y + dy[dir];
			int nx = x + dx[dir];
			if(OOB(ny,nx) ||  board[ny][nx] == '#')
				continue;
			if(board[ny][nx]=='S') {
				board[ny][nx] = '.';
				ret = Math.max(ret,  solve(ny,nx,cnt+1, t + 1));
				board[ny][nx] = 'S';
			}else {
				ret = Math.max(ret,  solve(ny,nx,cnt, t + 1));
			}
		}
		
		return ret;
	}
	
	static int dy[] = {-1,1,0,0};
	static int dx[] = {0,0,-1,1};
	static boolean OOB(int y,int x) {
		return y<=0 || y> R || x<=0 ||x > C;
	}
	
	static StringTokenizer st;
	static int R,C,T;
	static boolean visited[][];
	static char board[][];
}

File: 백준/Gold/21772. 가희의 고구마 먹방/가희의 고구마 먹방.py
from collections import namedtuple,deque


dy = [-1,1,0,0]
dx = [0,0,-1,1]

def OOB(y, x):
    return y<=0 or y > R or x<=0 or x> C


def solve(y,x, cnt, t):
    if t == T:
        return cnt

    # 현재 위치 그대로
    ret = solve(y,x, cnt ,t + 1)

    # 다음 칸 이동
    for d in range(4):
        ny = y + dy[d]
        nx = x + dx[d]

        if OOB(ny,nx) or board[ny][nx] == '#':
            continue

        # 다음칸이 S라면
        if board[ny][nx] == 'S':
            board[ny][nx] = '.'
            ret = max(ret, solve(ny,nx, cnt + 1, t + 1))
            board[ny][nx] = 'S'
        else:
            ret = max(ret, solve(ny,nx,cnt, t + 1))
        

    return ret


if __name__ == "__main__":
    R,C,T = map(int,input().split())
    visited = [[False for _ in range(C+1)] for _ in range(R+1)]
    Pair = namedtuple('Pair', ['y', 'x',])
    board = [['.'] for _ in range(R+1)]
    start = None
    for y in range(1,R+1):
        line = input()
        for x in range(1, C+1):
            board[y].append(line[x-1])
            if board[y][x] == 'G':
                start = Pair(y,x)

    visited[start.y][start.x] = True
    print(solve(start.y, start.x, 0, 0))

